mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,964 @@
+            break;
+          }
+        }
+
+        if (!onlyLastIndexVaries)
+          continue;
+
+        // Difference in values at last index should be less than
+        // cache_line_size/elementSize for a useful locality.
+        unsigned elementSize = this->elementSize[srcOp];
+        if (!(abs(srcOpCV[srcOpCV.size() - 1] -
+                  destOpCV[destOpCV.size() - 1]) <=
+              cache_line_size / elementSize))
+          continue;
+
+        // Insert `dstOp` into the group of `srcOp`.
+        insertIntoReferenceGroup(referenceGroups, groupId, srcOp, dstOp);
+
+      } else {
+        // Criteria 1
+        for (unsigned depth = 1; depth <= maxDepth + 1; depth++) {
+          FlatAffineConstraints dependenceConstraints;
+          SmallVector<DependenceComponent, 2> depComps;
+          DependenceResult result = checkMemrefAccessDependence(
+              srcAccess, dstAccess, depth, &dependenceConstraints, &depComps);
+
+          if (!hasDependence(result))
+            continue;
+          // Criteria 1.1
+          if (depth == maxDepth + 1) {
+            // Loop-independent dependence. Put both instructions in same
+            // reference group.
+            insertIntoReferenceGroup(referenceGroups, groupId, srcOp, dstOp);
+          } else {
+            // Criteria 1.2
+            // Search for dependence at depths other than innermostIndex.
+            // All entries should be zero.
+            bool hasDependency = false;
+            for (unsigned i = 0; i <= maxDepth; i++) {
+              if ((depComps[i].lb.getValue() != 0) &&
+                  (i != innermostIndex)) {
+                hasDependency = true;
+                break;
+              }
+            }
+            if (hasDependency)
+              continue;
+            else if (abs(depComps[innermostIndex]
+                             .lb.getValue()) <= 2)
+              insertIntoReferenceGroup(referenceGroups, groupId, srcOp, dstOp);
+          }
+        }
+      }
+    }
+  }
+  this->referenceGroups = referenceGroups;
+}
+
+/// Calculates number of cache lines accessed by each affine.for in the
+/// `loopNest` if each of those were considered the innermost loop.
+///
+/// Please refer Steve Carr et. al for detailed descriptions.
+/// https://dl.acm.org/doi/abs/10.1145/195470.195557
+void LoopInterchange::getCacheLineAccessCounts(
+    ArrayRef<AffineForOp> loopNest,
+    SmallVector<Operation *, 8> &loadAndStoreOps) {
+
+  unsigned loopNestSize = loopNest.size();
+  for (unsigned innerloop = 0; innerloop < loopNestSize; innerloop++) {
+    AffineForOp forOp = loopNest[innerloop];
+
+    // Build reference groups considering each `forOp` to be the inner most
+    // loop.
+    buildReferenceGroups(loadAndStoreOps, loopNestSize, innerloop + 1);
+
+    float step = forOp.getStep();
+    float trip =
+        (forOp.getConstantUpperBound() - forOp.getConstantLowerBound()) / step +
+        1;
+    long double totalCacheLineCount = 0;
```

Comment:
Why is a count a `double` entity insetad of uint64_t?

Reply:
Here, we try to generalize for the cases where `step` may not be 1. In that case, `step` should be kept float to avoid integer division in calculation of `trip`.  We can keep `stride` as uint64_t, but calculation of count entity `cacheLinesForThisOperation` requires `(trip*stride)/8` - again a double/float quantity. 

`totalCacheLineCount` is sum of `cacheLinesForThisOperation` ( a double) - hence this count entity is kept `double.`

For cases where we consider only `step` = 1 always, no double quantity is needed.

Reply:
Update: Replaced all double/float entities with appropriate unsigned values. For small values of `step`, this should have no impact. Thanks.